rotkehlchen/chain/ethereum/modules/curve/decoder.py

```diff
@@ -257,35 +259,155 @@ def _decode_curve_events(  # pylint: disable=no-self-use
+                event.counterparty == CPT_CURVE
+            ):
+                deposit_events.append(event)
+
+        # exit early if previously decoded deposit events aren't present.
+        if len(deposit_events) == 0:
+            log.debug(
+                f'Expected to see a receive pool token event and deposit events for a curve pool, '
+                f'but have not found them. Tx_hash: {transaction.tx_hash.hex()}',
+            )
+            return decoded_events
+
+        # search for the receipt of the wrapped token after deposit
+        for idx, event in enumerate(decoded_events):
+            if (
+                event.event_type == HistoryEventType.RECEIVE and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.location_label == transaction.from_address and
+                event.counterparty in CURVE_DEPOSIT_CONTRACTS
+            ):
+                receive_event_idx = idx
+                crypto_asset = event.asset.resolve_to_crypto_asset()
+                event.event_type = HistoryEventType.RECEIVE
+                event.event_subtype = HistoryEventSubType.RECEIVE_WRAPPED
+                event.counterparty = CPT_CURVE
+                event.notes = f'Receive {event.balance.amount} {crypto_asset.symbol} after depositing in curve pool'  # noqa: E501
+                break
+
+        # exit early if there's no event for the wrapped token receipt.
+        if receive_event_idx is None:
+            log.debug(
+                f'Expected to see a receive pool token event and deposit events for a curve pool, '
+                f'but have not found them. Tx_hash: {transaction.tx_hash.hex()}',
+            )
+            return decoded_events
+
+        decoded_events[receive_event_idx].extra_data = {'deposit_events_num': len(deposit_events)}  # for accounting  # noqa: E501
+        previous_event = decoded_events[receive_event_idx]
+        for event in deposit_events:
+            maybe_reshuffle_events(previous_event, event, decoded_events)
+            previous_event = event
+
+        return decoded_events
+
+    def _handle_zap_contracts_liquidity_removal(
+            self,
+            transaction: EvmTransaction,
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+    ) -> list[HistoryBaseEntry]:
+        """
+        This method completes the decoding of liquidity removal made through
+        Curve zap contracts such as e.g. Curve Deposit USDN, etc.
+        """
+        return_event_idx, withdrawal_events = None, []
+        for idx, event in enumerate(decoded_events):
+            if (
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.RETURN_WRAPPED and
+                event.counterparty == CPT_CURVE
+            ):
+                # exit early since transaction was decoded properly.
```

Comment:
But isn't it guaranteed that if it is a zap deposit contract then it will not be decoded correctly?

Reply:
Also if I am not mistaken then currently if a transaction has multiple deposits / withdrawals to pools then if at least one of the deposits / withdrawals is decoded correctly the post decoding rules won't run. Am I correct? If yes, then this looks like something that needs to change.

Reply:
> But isn't it guaranteed that if it is a zap deposit contract then it will not be decoded correctly?

Yeah, until it gets to the post-decoding process.

> Also if I am not mistaken then currently if a transaction has multiple deposits / withdrawals to pools then if at least one of the deposits / withdrawals is decoded correctly the post decoding rules won't run. Am I correct? If yes, then this looks like something that needs to change.

Not sure I totally understand you here but I'd describe what I understood...
## 1.
>If a transaction has multiple deposits/withdrawals to pools

If this happens, then I expect(not entirely sure) the already existing decoder to handle it. The post decoding only exists to handle decoding liquidity addition/removal via the **Zap depositer contracts**.

Also, looking at the transactions page of the contracts(Curve Deposit: USDN), they all happen to call a single function(you're either removing liquidity or adding it). I didn't come across a transaction where an aggregation(multicall) of method calls occur(scanning through the ~25 pages of transactions).

Sorry but if you could point me to a sample txn where such scenario occur, to avoid a revisit on this at a later point I'd really appreciate it.

## 2.
> then if at least one of the deposits / withdrawals is decoded correctly the post decoding rules won't run. Am I correct?

Yeah, you're correct that if a combination of `AddLiquidity` & `RemoveLiquidity` events in a single txn for Zap depositer contracts occur. It'd be an unexpected behaviour.

Reply:
Yeah, I see. I think we then can consider a limitation the fact that it won't work good with multiple deposits / withdrawals in one transaction.

Example transaction with multiple deposits btw https://etherscan.io/tx/0xc2c0c5e2941795132bc06b7933b130471d6d147a6ae9f0a7beadca4532fb93b1/

Reply:
Alexey is right and I think this is the same thing as I mentioned in `_handle_post_decoding` function. I don't want to put more limitations when we know about them. It should work for all cases, so this needs to be changed.

Ofcourse there can be multiple deposits and withdrawals in different pools in one transaction. That's the whole point of a Zap contract. To combine stuff. Lots of crazy combinations in a single transaction exist out there.

you need to write decoders that are future-proof and work for all of that. not break easily.